### DISCUSSION TITLE
Did not invoke after hook.

### DISCUSSION THREAD

**Discussion by re2zero** [bytedance/android-inline-hook#8] [Q&A] [answered]
```cpp
bool InputDispatcher::hasInjectionPermission(int32_t injectorPid, int32_t injectorUid)
```
hook 成功找到函数地址了，但是没有被执行，求大佬给看看 
```cpp
typedef bool (*type_t)(int32_t injectorPid, int32_t injectorUid);
//static bool (*orgin_hasInjectionPermission)(int32_t injectorPid, int32_t injectorUid);

// 代替的方法
static bool new_hasInjectionPermission(int32_t injectorPid, int32_t injectorUid) {
  LOG("HAHSHDAHHDA->>>>>>> %d, %d\n", injectorPid, injectorUid);
  ((type_t)orig)(injectorPid, injectorUid);
  return true;
}
...
 stub = shadowhook_hook_sym_name(
          "libinputflinger.so",
          "_ZN7android15InputDispatcher22hasInjectionPermissionEii",
          (void *)new_hasInjectionPermission,
          (void **)&orig);

#log:
I/shadowhook_tag: shadowhook: hook_sym_name(libinputflinger.so, _ZN7android15InputDispatcher22hasInjectionPermissionEii, 0x73322f8100) OK. return: 0x738e5686c0. 1 - Pending task
```

**Comment by caikelun** [comment, marked as answer]
hook 的时候 `libinputflinger.so` 还没有被加载。errno 为 `1`，详见文档：
https://github.com/bytedance/android-inline-hook/blob/main/doc/manual.zh-CN.md#%E9%94%99%E8%AF%AF%E7%A0%81

**Reply by re2zero** [reply]
@caikelun 非常感谢。 再请教下，这个库属于系统服务加载，也就是hook的函数要调用成功，只能是自己App加载的动态库吗？

**Reply by caikelun** [reply]
系统加载的动态库和app加载的动态库都可以hook，没有区别。
对于这个库，你可以等到系统加载了它再执行hook，或者你也可以自己手动提前加载它。

**Reply by re2zero** [reply]
输入事件，系统肯定加载了，原函数里的日志都打印出来了，但是hook依然是1，也没有进去。有没有系统服务动态库的示例呀
想用hook方法，分发输入事件到系统，看来要放弃*_*

**Reply by caikelun** [reply]
你可以从 maps 确认下这个库是否被加载了。

**Reply by re2zero** [reply]
系统服务启动加载libandroid_servers.so,但是maps也看不到此库被加载，放弃。

**Reply by xmutzlq** [reply]
为什么我用readelf -s libandroid.so是这样的结果：
189: 00003001     2 FUNC    GLOBAL DEFAULT    6 AAssetDir_close
   190: 00003003     2 FUNC    GLOBAL DEFAULT    6 AAssetDir_getNextFileName
   191: 00003005     2 FUNC    GLOBAL DEFAULT    6 AAssetDir_rewind
   192: 00003007     2 FUNC    GLOBAL DEFAULT    6 AAssetManager_fromJava
   193: 00003009     2 FUNC    GLOBAL DEFAULT    6 AAssetManager_open
   194: 0000300b     2 FUNC    GLOBAL DEFAULT    6 AAssetManager_openDir
   195: 0000300d     2 FUNC    GLOBAL DEFAULT    6 AAsset_close
   196: 0000300f     2 FUNC    GLOBAL DEFAULT    6 AAsset_getBuffer
   197: 00003011     2 FUNC    GLOBAL DEFAULT    6 AAsset_getLength
   198: 00003013     2 FUNC    GLOBAL DEFAULT    6 AAsset_getLength64
   199: 00003015     2 FUNC    GLOBAL DEFAULT    6 AAsset_getRemainingLength
   200: 00003017     2 FUNC    GLOBAL DEFAULT    6 AAsset_getRemainingLength
   201: 00003019     2 FUNC    GLOBAL DEFAULT    6 AAsset_isAllocated
   202: 0000301b     2 FUNC    GLOBAL DEFAULT    6 AAsset_openFileDescriptor
   203: 0000301d     2 FUNC    GLOBAL DEFAULT    6 AAsset_openFileDescriptor
   204: 0000301f     2 FUNC    GLOBAL DEFAULT    6 AAsset_read
   205: 00003021     2 FUNC    GLOBAL DEFAULT    6 AAsset_seek
   206: 00003023     2 FUNC    GLOBAL DEFAULT    6 AAsset_seek64
   hook成功，确没有回调
   I/shadowhook_tag: shadowhook: hook_sym_name(libandroid.so, AAssetManager_open, 0x79f046944c) OK. return: 0x79ff626620. 0 - OK
`typedef void (*type_t)(AAssetManager *, const char *, int);

static void proxy(AAssetManager *assetManager, const char *filename, int mode)
{
  // do something
  LOG("do something %s", "before");
  ((type_t)orig)(assetManager, filename, mode);
  LOG("do something %s", "after");
  // do something
}
stub = shadowhook_hook_sym_name(
          "libandroid.so",
          "AAssetManager_open",
          (void *)proxy,
          (void **)&orig);
`
